python/rascaline/rascaline/utils/clebsch_gordan/_density_correlations.py

```diff
@@ -0,0 +1,258 @@
+"""
+This module provides convenience calculators for preforming density correlations, i.e.
+the (iterative) CG tensor products of density (body order 2) tensors.
+
+All of these calculators wrap the :py:class:`TensorCorrelator` class, handling the
+higher-level metadata manipulation to produce the desired output tensors.
+"""
+
+from typing import List, Optional, Tuple
+
+from .. import _dispatch
+from .._backend import Labels, TensorMap, TorchModule, torch_jit_export
+from . import _utils
+from ._tensor_correlator import TensorCorrelator
+
+
+class DensityCorrelations(TorchModule):
+    """
+    Iterative products of a density to form higher arbitrary body order tensors.
+
+
+    :param n_correlations: :py:class:`int`, the number of iterative CG tensor products
+        to perform.
+    :param angular_cutoff: :py:class:`int`, the maximum angular momentum of output
+        blocks to compute at any CG tensor product iteration. If ``None``, the maximum
+        angular momentum is determined by the combination of the angular order of the
+        tensor input to :py:meth:`compute` and the number of iterations.
+    :param skip_redundant: :py:class:`bool`, whether to skip redundant computations on
+        intermediate iterations. This parameter should only be set to ``True`` if
+        performing density auto-correlations and should not be used otherwise.
+    :param tensor_correlator: :py:class:`TensorCorrelator`, the calculator to use for
+        the CG tensor product. If ``None``, a new :py:class:`TensorCorrelator` is
+        initialized with the provided or default settings. Passing this arguments saves
+        re-computation of CG coefficients. If passed, the ``max_angular`` used to
+        initialize it must be high enough to handle all tensor products performed by
+        this calculator.
```

Comment:
Ok so I agree that for 'convenience calculators' (i.e. `EquivariantPowerSpectrum`) we should not have leaky abstraction. My consideration is for model developers trying to easily prototype new descriptors. 

In these cases, I think TensorCorrelator should be initialized once and passed around, because if you do enough tensor products it does add cost onto the process (I did non-rigorous speed tests for passing vs not passing in a notebook and it wasn't negligible). Especially as for a torch-scriptable NN you'll need to pre-initialize every layer anyway at the same time, I think it makes sense not to unnecessarily increase constructor overhead and memory.

So next the question is whether `DensityCorrelations` is a convenience calculator or not, and I'd argue it isn't. In some complicated use cases Jigyasa mentioned, a n-body correlation is just one of the tensor in a long line of tensors to be correlated. Having a `DensityCorrelations` calculator to generate this easily without having to worry about the metadata is useful, and in this model developer case I'd want to pass the `TensorCorrelator` around here too.

Reply:
I don't understand. Either you do 

```py
tc = TensorCorrelator()
dc = DensityCorrelations(tc)
```

or you do

```py
dc = DensityCorrelations()
```

In both cases the `TensorCorrelator` is only created once. If you want to keep your own `TensorCorrelator` outside of the `DensityCorrelations`, you can do this with

```py
dc = DensityCorrelations()

tc = TensorCorrelator()
```

And then we only initialize two `TensorCorrelator` which is going to be negligible against the actual calculations. 

From here you'll want to re-use `tc` and/or `dc` across calculations for sure, but that's unrelated. The one thing you don't want to do is re-create new calculators at each calculation step, but creating one or two at the start should be completely fine.

Reply:
Noted - removed the option of passing `tensor_correlator` to `DensityCorrelations`